### PR TITLE
[ASL] Start Transliterating Typing Rules

### PR DESCRIPTION
**Pull request by jalglave**
hi @HadrienRenaud please could you have a look at the phrasing and comment as you see fit?
Thanks,
Jade

### CODE REVIEW

File: asllib/ASLTypingReference.mld [hunk 1]
```diff
@@ -28,38 +28,113 @@ annotate a specific object, as follows.
...
+      - [t] is [t'];
+      - [new_e] is [e'']. 
+    - All of the following applies:
+      - [t''] is a structural subtype of [t'] in [env];
+      - [t''] is not a domain subtype of [t'] in [env];
+      - an execution-time check that the expression evaluates to a value in the
+        domain of the required type is required.
+   - All of the following applies:
+     - [t''] is not a structural subtype of [t'] in [env];
+     - [t''] is not a domain subtype of [t'] in [env];
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:64]: This line is not required (as not structural subtypes => not domain subtype)

```suggestion
```

File: asllib/ASLTypingReference.mld [hunk 2]
```diff
@@ -28,38 +28,113 @@ annotate a specific object, as follows.
...
+
+{3 TypingRule.ELocalVarConstant}
+
+  {4 Rule}
+  All of the following applies:
+  - [e] is a variable [x];
+  - [x] maps to a type [ty] in the [storage_types] of the local environment given by [env];
+  - [x] maps to a local constant [v];
+  - [t] is [ty];
+  - [new_e] is [E_Literal v].
```
Comment (HadrienRenaud) [inline asllib/ASLTypingReference.mld:77]: I don't think that we've been using the `AST.mli` constructors to deconstruct the AST (see for example at L73: "[e] is a variable [x]"), so I think we should say
```suggestion
  - [new_e] is the Literal [v].
```

**Review by HadrienRenaud — COMMENTED** [review]
Looks very nice, I'll write some examples when I have a bit of time.

**Review by relokin — APPROVED** [review]
This change looks good to me and all unit tests run successfully. I'll go ahead and merge shortly.

**Pull request merged**